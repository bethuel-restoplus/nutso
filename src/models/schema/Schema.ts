import { ArraySchema } from "./ArraySchema";
import { BooleanSchema } from "./BooleanSchema";
import { DateSchema } from "./DateSchema";
import { NumberSchema } from "./NumberSchema";
import { ObjectSchema } from "./ObjectSchema";
import { StringSchema } from "./StringSchema";
import { TupleSchema } from "./TupleSchema";

// prettier-ignore
export type Schema<T, R = T> = T extends string ? StringSchema<R> :
                        T extends number ? NumberSchema :
                        T extends Date ? DateSchema :
                        T extends boolean ? BooleanSchema :
                        T extends Array<infer E> ? ArraySchema<E, R> : 
                        T extends [infer U, ...unknown[]] ? TupleSchema<U> :
                        T extends Object ? ObjectSchema<T, R> :
                        never;
