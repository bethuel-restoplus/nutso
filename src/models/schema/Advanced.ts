type False = "0";
type True = "1";
type If<C extends True | False, Then, Else> = { "0": Else; "1": Then }[C];

type Diff<T extends string | number | symbol, U extends string> = ({ [P in T]: P } & { [P in U]: never } & { [x: string]: never })[T];

type X<T> = Diff<keyof T, keyof Object>;

type Is<T, U> = (Record<X<T & U>, False> & Record<any, True>)[Diff<X<T>, X<U>>];

type DeepPartial<T> = {
  [P in keyof T]?: If<Is<Function & T[P], Function>, T[P], DeepPartial<T[P]>>;
};

type Customer = {
  name: string;
  address: {
    line1: string;
    city: string;
    props: { [key: string]: string };
  };
};

type Props = { [key: string]: string };
const props: Props = { 1: "a" };

type C1 = DeepPartial<Customer>;

const c1: C1 = {
  name: "John",
  address: {
    line1: "66 Collins St",
    props: {
      1: "foo"
    }
  }
};

type Foo = {
  type: "foo";
  foo: string;
};

type Bar = {
  type: "bar";
  bar: number;
};

type Baz = Foo | Bar;
const baz: Baz = {
  type: "foo",
  foo: "string"
};
