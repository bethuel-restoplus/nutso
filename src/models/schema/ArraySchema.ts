import { OptionalFlag } from "../OptionalFlag";
import { Schema } from "./Schema";

/**
 * E => Element
 * T => Type (this is the array)
 * P => Parent of the array
 *
 * Look at this type, we need all 3 types
 */
export type ArraySchema<E, T extends E[], P> = {
  type: "array";
  optional?: OptionalFlag<P>;
  minItems?: number;
  maxItems?: number;
  items: Schema<E, T>;
};
