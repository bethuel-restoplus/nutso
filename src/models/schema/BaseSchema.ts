// export type BaseSchema<R> = {
//   optional?: OptionalFlag<R>;
// };
