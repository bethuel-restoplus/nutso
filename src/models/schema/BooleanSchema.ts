import { OptionalFlag } from "../OptionalFlag";

export type BooleanSchema<P = unknown> = {
  type: "boolean";
  optional?: OptionalFlag<P>;
};
