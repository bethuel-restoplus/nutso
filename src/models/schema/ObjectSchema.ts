import { Schema } from "./Schema";
import { OptionalFlag } from "../OptionalFlag";

export type ObjectSchema<T, P = unknown> = {
  type: "object";
  optional?: OptionalFlag<P>;
  properties: {
    [K in keyof T]: Schema<T[K], T>;
  };
};
