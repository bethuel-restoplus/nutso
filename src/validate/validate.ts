import { Schema } from "../models/schema/Schema";
import { Result } from "../models/result/Result";
import { validateString } from "./validateString";
import { StringSchema } from "../models/schema/StringSchema";
import { NumberSchema } from "../models/schema/NumberSchema";
import { validateNumber } from "./validateNumber";
import { BooleanSchema } from "../models/schema/BooleanSchema";
import { validateBoolean } from "./validateBoolean";
import { ObjectSchema } from "../models/schema/ObjectSchema";
import { validateObject } from "./validateObject";
import { FieldPath } from "../models/FieldPath";
import { validateArray } from "./validateArray";
import { ArraySchema } from "../models/schema/ArraySchema";
import { validateDate } from "./validateDate";
import { DateSchema } from "../models/schema/DateSchema";

/**
 * This is an internal API that takes the root as an argument
 * and called recursively from `validateObject` and `validateArray` methods
 *
 * @param o
 * @param root
 * @param schema
 */
export const _validate = <T, R>(o: T | null, root: R, schema: Schema<T, R>): Result<T> => {
  //
  if (!schema) {
    throw new Error("Schema should not be null");
  }

  /**
   * validate object against the schema
   */
  switch (schema.type) {
    case "number":
      return validateNumber(o, root, schema as NumberSchema<R>) as Result<T>;
    case "string":
      return validateString(o, root, schema as StringSchema<R>) as Result<T>;
    case "boolean":
      return validateBoolean(o, root, schema as BooleanSchema<R>) as Result<T>;
    case "array":
      return validateArray(o as any, root, schema as ArraySchema<T, R>) as Result<T>;
    case "object":
      return validateObject(o, root, schema as ObjectSchema<T, R>) as Result<T>;
    case "date":
      return validateDate(o, root, schema as DateSchema<R>) as Result<T>;
  }
  throw new Error(`Unhandled data type`);
};

export const validate = <T>(o: T, schema: Schema<T>): Result<T> => {
  return _validate(o, o, schema);
};
