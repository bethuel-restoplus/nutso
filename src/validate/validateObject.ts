import { ObjectResult } from "../models/result/ObjectResult";
import { ObjectSchema } from "../models/schema/ObjectSchema";
import { isNil } from "../utils/typeChecker";
import { _validate } from "./validate";

export const isRegex = (str: string) => {
  return str.startsWith("/") && str.endsWith("/");
};

export const getRegex = (str: string) => {
  return new RegExp(`^${str.substr(1, str.length - 1)}$`);
};

export const validateObject = <T extends { [key: string]: any }, R>(
  o: T | null,
  root: R,
  schema: ObjectSchema<T, R>
): ObjectResult<T> => {
  //
  const result: ObjectResult<T> = {
    isValid: true,
    errorMessage: "",
    properties: {} as any,
    errorPath: []
  };

  // isnil - no further traversal required
  if (isNil(o)) {
    // valid
    if (schema.optional) {
      result.isValid = true;
      return result;
    }
    // invalid
    result.isValid = false;
    result.errorMessage = `Required field.`;
    return result;
  }

  const obj = o as T;
  const processedFields: string[] = [];

  // for each static key, validate
  for (let field in schema.properties) {
    if (isRegex(field)) continue;
    result.properties[field] = _validate(o ? o[field] : null, root, schema.properties[field]);
    processedFields.push(field);
  }

  // for each regex in schema.properties
  for (let field in schema.properties) {
    if (!isRegex(field)) continue; // not-regex
    const regex = getRegex(field);

    // for each key
    for (let key in obj) {
      /**
       * Already processed field. This will happen if the regex matches a wide range of fields, like ".*"
       * and the properties also defines specific fields like "red": {}, "blue": {} etc ...
       */
      if (processedFields.includes(key)) continue;
      if (!regex.test(key)) continue;
      result.properties[key] = _validate(o ? o[key] : null, root, schema.properties[field]);
      processedFields.push(key);
    }
  }

  // if this node is valid, then check if all of it's children are valid
  // because the node is invalid, if any of it's children are invalid
  if (result.isValid) {
    for (let field of processedFields) {
      const property = result.properties[field];
      if (!property.isValid) {
        result.isValid = false;
        result.errorMessage = property.errorMessage;
        result.errorPath = [field, ...property.errorPath];
        break;
      }
    }
  }

  return result;
};
