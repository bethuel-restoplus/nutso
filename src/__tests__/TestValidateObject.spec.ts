import { ObjectSchema } from "../models/schema/ObjectSchema";
import { Schema } from "../models/schema/Schema";
import { validate, _validate } from "../validate/validate";
import { validateObject } from "../validate/validateObject";

test(`Basic`, () => {
  type Customer = {
    name: "";
  };

  const obj: Customer = { name: "" };
  const schema: ObjectSchema<Customer, Customer> = {
    type: "object",
    properties: {
      name: {
        type: "string",
        minLength: 3
      }
    }
  };

  const result1 = validateObject(obj, obj, schema);
  expect(result1.isValid).toBe(false);

  const result2 = _validate(obj, obj, schema);
  expect(result2.isValid).toBe(false);
});

test(`Address is optional in schema and undefined in object`, () => {
  type Customer = {
    name: string;
    address?: {
      city: string;
    };
  };

  const obj: Customer = { name: "John" };
  const schema: Schema<Customer> = {
    type: "object",
    properties: {
      name: {
        type: "string",
        minLength: 3
      },
      address: {
        type: "object",
        optional: true,
        properties: {
          city: {
            type: "string"
          }
        }
      }
    }
  };

  const result = validate(obj, schema);
  expect(result).toMatchSnapshot();
});

test(``, () => {
  type Customer = {
    name: string;
    props: { [key: string]: string };
  };

  const schema: Schema<Customer> = {
    type: "object",
    properties: {
      name: {
        type: "string"
      },
      props: {
        type: "object",
        properties: {}
      }
    }
  };
});
