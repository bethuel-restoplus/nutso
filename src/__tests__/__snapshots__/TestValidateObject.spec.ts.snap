// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Address is optional in schema and undefined in object 1`] = `
Object {
  "errorMessage": "",
  "errorPath": Array [],
  "isValid": true,
  "properties": Object {
    "address": Object {
      "errorMessage": "",
      "errorPath": Array [],
      "isValid": true,
      "properties": Object {},
    },
    "name": Object {
      "errorMessage": "",
      "errorPath": Array [],
      "isValid": true,
    },
  },
}
`;
